source/slang/type-layout.cpp

```diff
@@ -125,86 +149,179 @@ struct DefaultLayoutRulesImpl : SimpleLayoutRulesImpl
+    /// Common behavior for GLSL-family layout.
+struct GLSLBaseLayoutRulesImpl : DefaultLayoutRulesImpl
+{
+    typedef DefaultLayoutRulesImpl Base;
+
+    SimpleLayoutInfo GetVectorLayout(SimpleLayoutInfo elementInfo, size_t elementCount) override
+    {
+        // The `std140` and `std430` rules require vectors to be aligned to the next power of
+        // two up from their size (so a `float2` is 8-byte aligned, and a `float3` is
+        // 16-byte aligned).
+        //
+        // Note that in this case we have a type layout where the size is *not* a multiple
+        // of the alignment, so it should be possible to pack a scalar after a `float3`.
+        //
+        SLANG_RELEASE_ASSERT(elementInfo.kind == LayoutResourceKind::Uniform);
+        SLANG_RELEASE_ASSERT(elementInfo.size.isFinite());
+
+        auto size = elementInfo.size.getFiniteValue() * elementCount;
+        SimpleLayoutInfo vectorInfo(
+            LayoutResourceKind::Uniform,
+            size,
+            RoundUpToPowerOfTwo(size));
+        return vectorInfo;
+    }
+
+    SimpleArrayLayoutInfo GetArrayLayout( SimpleLayoutInfo elementInfo, LayoutSize elementCount) override
+    {
+        // The size of an array must be rounded up to be a multiple of its alignment.
+        //
+        auto info = Base::GetArrayLayout(elementInfo, elementCount);
+        info.size = RoundToAlignment(info.size, info.alignment);
+        return info;
+    }
+
+    void EndStructLayout(UniformLayoutInfo* ioStructInfo) override
+    {
+        // The size of a `struct` must be rounded up to be a multiple of its alignment.
+        //
         ioStructInfo->size = RoundToAlignment(ioStructInfo->size, ioStructInfo->alignment);
     }
 };
 
-// Capture common behavior betwen HLSL and GLSL (`std140`) constnat buffer rules
-struct DefaultConstantBufferLayoutRulesImpl : DefaultLayoutRulesImpl
+    /// The GLSL `std430` layout rules.
+struct Std430LayoutRulesImpl : GLSLBaseLayoutRulesImpl
 {
-    // The `std140` rules require that all array elements
-    // be a multiple of 16 bytes.
-    //
-    // HLSL agrees.
+    // These rules don't actually need any differences from our
+    // base/common GLSL layout rules.
+};
+
+    /// The GLSL `std430` layout rules.
+struct Std140LayoutRulesImpl : GLSLBaseLayoutRulesImpl
+{
+    typedef GLSLBaseLayoutRulesImpl Base;
+
     SimpleArrayLayoutInfo GetArrayLayout(SimpleLayoutInfo elementInfo, LayoutSize elementCount) override
     {
+        // The `std140` rules require that array elements
+        // be aligned on 16-byte boundaries.
+        //
         if(elementInfo.kind == LayoutResourceKind::Uniform)
         {
             if (elementInfo.alignment < 16)
                 elementInfo.alignment = 16;
-            elementInfo.size = RoundToAlignment(elementInfo.size, elementInfo.alignment);
         }
-        return DefaultLayoutRulesImpl::GetArrayLayout(elementInfo, elementCount);
+        return Base::GetArrayLayout(elementInfo, elementCount);
```

Comment:
I've used the convention (on a fair amount of code) of using a 'Parent' typedef for this usage.

Reply:
That seems reasonable, although in the extreme a single object type may have all three of:

* A “base” type in the inheritance hierarchy (ignoring multiple inheritance...)
* A “parent” object in some kind of run-time hierarchy
* An “outer” type that it is lexically nested inside, and maybe an outer object of that type which represents runtime context

I don’t know if those exact terms are the best for the three concepts, but I’d argue that in the limit it can help to be able to keep them conceptually distinct. (I have similar opinions about people using “dimension” in places where they mean one or more of shape/rank/dimensionality, extents/size, or an axis. :)

Reply:
The 3 examples you give could be useful (especially if concepts such as parent and base made simple sense) meant a language just defined them - so you didn't need to book keep. The other name that I think swift (and some others) use is 'Super'. 
Anyway in practice the most common used of the 3 by far is Parent/Super in my experience.

Reply:
I like `Super` then, because it is unambiguous.

I’ll plan to clean up this and the other style comment before I check in.